### PR TITLE
Add eslint config scaffold

### PR DESCRIPTION
**Pull request by RusinovAnton**
Add `.eslintrc` that extends `airbnb/base` without any plugins.

refs #1 
I created branch [test-eslint](https://github.com/yola/sitebuilderui/compare/test-eslint) in sitebuilderui to test eslint config script. To test locally:
- Clone eslint-config-yola and link package:  
`git clone -b rusinov/init-repo git@github.com:yola/eslint-config-yola.git && cd eslint-config-yola && npm link`
- Go to `sitebuilderui` and checkout branch `test-eslint`, then run:  
`npm i && npm link eslint-config-yola && npm run eslint`

### CODE REVIEW

**Comment by RusinovAnton** [comment]
@yola/frontend-engineering

**Comment by eugeneford** [comment]
@RusinovAnton, how do you think, shall we add ``index.js`` as was done here: https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb/index.js?

**Comment by RusinovAnton** [comment]
While testing current setup in https://github.com/yola/sitebuilderui/compare/test-eslint got this error
```
> sitebuilderui@0.0.0 eslint /home/rusinov/yola/sitebuilderui
> eslint src/apps/**/*.js


Oops! Something went wrong! :(

ESLint: 4.10.0.
ESLint couldn't find the plugin "eslint-plugin-import". This can happen for a couple different reasons:

1. If ESLint is installed globally, then make sure eslint-plugin-import is also installed globally. A globally-installed ESLint cannot find a locally-installed plugin.

2. If ESLint is installed locally, then it's likely that the plugin isn't installed correctly. Try reinstalling by running the following:

    npm i eslint-plugin-import@latest --save-dev

If you still can't figure out the problem, please stop by https://gitter.im/eslint/eslint to chat with the team.

```

**Comment by beck** [comment]
Reading [the documentation](https://github.com/airbnb/javascript/blob/16faade698a185d3a84f4c2c8ef70ec4cf712936/packages/eslint-config-airbnb-base/README.md) it says:

> requires `eslint` and `eslint-plugin-import`

So all of our FE apps will need to install:
* `eslint`
* `eslint-plugin-import`
* `eslint-config-yola`

That's not too bad.

**Comment by RusinovAnton** [comment]
> So all of our FE apps will need to install ... 
Yeah, actually, I like that approach more than one in which we would have to write wrapper plugin to extract rules from plugins and extend it then.

**Comment by beck** [comment]
This needs a readme update.

The readme should note:
* the style guide we're using as our base
* how to install this into a FE app (and note why `eslint-plugin-import` is needed)

File: package.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/yola/eslint-config-yola/issues"
+  },
+  "homepage": "https://github.com/yola/eslint-config-yola#readme",
+  "dependencies": {
+    "eslint-config-airbnb-base": "^12.1.0"
+  },
+  "peerDependencies": {
+    "eslint": "^4.10.0"
```
Comment (beck) [inline package.json]: Keep dependencies of our installable loose, use `4.x`

File: package.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    "index.js"
+  ],
+  "author": "Yola Engineering <engineers@yola.com>",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/yola/eslint-config-yola/issues"
+  },
+  "homepage": "https://github.com/yola/eslint-config-yola#readme",
+  "dependencies": {
+    "eslint-config-airbnb-base": "^12.1.0"
```
Comment (beck) [inline package.json]: Same here, lets use `12.x`

File: index.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  "extends": "airbnb-base"
```
Comment (beck) [inline index.js]: This fails for me.

It is not just plug-ins, but all peer deps are going to fail.

Reading through the issue I linked in the parent, one fix here is to use:

```
'extends': require.resolve('eslint-config-airbnb-base'),
```

This is a bit of a hack but will hold us over until eslint gets their act together.

Comment (beck) [inline index.js]: This is a javascript file, not json. Use single quotes and add a trailing comma.

**Comment by beck** [comment]
This is missing a `.eslintrc` that extends `index.js`. We want the javascript in this repo to also be linted

File: index.js (hunk 3 above)
Comment (RusinovAnton) [reply index.js]: How do you test that? That works for me in `test-eslint` just have to install dependencies manually after `npm link eslint-config-yola`

File: package.json (hunk 2 above)
Comment (RusinovAnton) [reply package.json]: I think that we should move it to `peerDependencies` if this idea https://github.com/yola/eslint-config-yola/issues/1#issuecomment-340464368 wouldn't work.

**Comment by RusinovAnton** [comment]
> how to install this into a FE app (and note why eslint-plugin-import is needed)
I would like to test it with published library before writing that guide.

**Comment by beck** [comment]
> I would like to test it with published library before writing that guide.

It should be possible to test it now and write documentation.

Some folks like to write their documentation _first_ before touching any code, they call it DDD - Document Driven Development.

File: package.json (hunk 2 above)
Comment (beck) [reply package.json]: That would then make it then four deps we'd have to install in every repo to add eslinting.
* two is great (the linter and config)
* three is annoying
* four makes me ask "how do we fix this?"

File: README.md [hunk 4]
```diff
@@ -1 +1,20 @@
...
+
+Shareable eslint configuration that ensures code style quality for Yola front-end codebase.  
+
+Code style of modern ES6 code is based on [Airbnb Javascript Style Guide](https://github.com/airbnb/javascript).
+
+## Setup
+To setup eslint for project you need to:
+- install eslint:  
+```
+npm install --save-dev eslint@^4.x
```
Comment (beck) [inline README.md]: Carets and `x` are distinct ways to specify version ranges, it does not make sense to use both.

File: README.md [hunk 5]
```diff
@@ -1 +1,20 @@
...
+``` 
+- install config package:  
+```
+npm install --save-dev eslint-config-yola@latest
```
Comment (beck) [inline README.md]: by default `@latest` is used, no? If yes, that would make it redundant.

File: README.md [hunk 6]
```diff
@@ -1 +1,20 @@
...
+```
+- create `.eslintrc` file and declare config to extend:
```
Comment (beck) [inline README.md]: One thing really nice about eslint is that it supports multiple config file types:
https://eslint.org/docs/user-guide/configuring#configuration-file-formats

Lets use `.eslintrc.js`, that gives us the ability to add comments if the app needs to add overriding config

File: package.json [hunk 7]
```diff
@@ -1,6 +1,9 @@
 {
   "name": "eslint-config-yola",
   "version": "0.1.0",
+  "scripts": {
+    "eslint": "eslint ./**/*.js"
```
Comment (RusinovAnton) [inline package.json]: Is it worth to add prepublish hook for repo?
```
"prepublish": "npm run lint",
```

File: package.json [hunk 8]
```diff
@@ -2,7 +2,7 @@
   "name": "eslint-config-yola",
   "version": "0.1.0",
   "scripts": {
-    "eslint": "eslint ./**/*.js"
+    "lint": "eslint -c .eslintrc.js ./**/*.js"
```
Comment (RusinovAnton) [inline package.json]: Is it worth to add prepublish hook for repo?
```
"prepublish": "npm run lint",
```

**Comment by RusinovAnton** [comment]
@yola/frontend-engineering 
should I define `react` and `jsx` eslint plugins as dependencies in current PR?

File: package-lock.json [hunk 9]
```diff
@@ -0,0 +1,1445 @@
+{
```
Comment (stefanor) [inline package-lock.json]: We're planning to use this as a library, so it probably shouldn't have a lock file.

Comment (RusinovAnton) [reply package-lock.json]: It isn't clear for me why not? Could you please explain what do you mean?

Comment (stefanor) [reply package-lock.json]: Are you intending anything that consumes this library to inherit all these pins? What if they're incompatible with the consumer's pins?

Comment (RusinovAnton) [reply package-lock.json]: Unlike `npm-shrinkwrap.json`, `package-lock.json` isn't being included in package and serve for library development only.
refs: https://docs.npmjs.com/files/package-lock.json#description

Comment (stefanor) [reply package-lock.json]: What is the point of that, for a library?

Comment (beck) [reply package-lock.json]: Lets move the package lock conversation to:
https://github.com/yola/production/issues/3850

TLDR; I say 🔥

Comment (RusinovAnton) [reply package-lock.json]: My bad, never seen that discussion before.

Comment (beck) [reply package-lock.json]: > My bad

No need to apologize, it is an old ticket and never had any solid resolution. So I'm making a decision: 🔥 !

Comment (RusinovAnton) [reply package-lock.json]: done 🎉

**Comment by beck** [comment]
> should I define react and jsx eslint plugins as dependencies in current PR?

Yeah, do it. Make sure we have this whole mess with sub-dependencies and plugins figured out.

File: package.json [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  ],
+  "author": "Yola Engineering <engineers@yola.com>",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/yola/eslint-config-yola/issues"
+  },
+  "homepage": "https://github.com/yola/eslint-config-yola#readme",
+  "dependencies": {
+    "eslint-config-airbnb-base": "12.x",
+    "eslint-plugin-import": "2.x"
```
Comment (beck) [inline package.json]: I thought we concluded this needed to be a peer-dep? Or was that fixed?

File: .npmrc [hunk 11]
```diff
@@ -0,0 +1 @@
+package-lock = false
```
Comment (beck) [inline .npmrc]: configure you editor to always add a newline at the end of the file

File: package.json (hunk 10 above)
Comment (RusinovAnton) [reply package.json]: I've tested that and it worked for me 
https://github.com/yola/eslint-config-yola/issues/1#issuecomment-340455449, https://github.com/RusinovAnton/eslint-config-rusinov

**Comment by beck** [comment]
Pulling up the conversation here because it has been collapsed:

> How do you test that? 

Same way I test any other lib.

From my `eslint-config-yola` checkout I run:

```
npm ln
```

Then I jump over to another app, in this case SBUI, and do:
```
npm ln eslint-config-yola
npm install eslint
touch .eslintrc.js
eslint src/ide/modules/index.js
# no output, all good
```

update `.eslintrc.js` to be:
```js
module.exports = {
  extends: 'yola',
}
```

Run:
```
 $ eslint src/ide/modules/index.js
Cannot find module 'eslint-config-airbnb-base'
Referenced from: /Users/doug/Yola/eslint-config-yola/index.js
Referenced from: /Users/doug/Yola/sitebuilderui/.eslintrc.js
Error: Cannot find module 'eslint-config-airbnb-base'
```

Which makes sense, because the eslint maintainers say "make airbnb a peer dependency" but we'd like to avoid doing that.

**Comment by RusinovAnton** [comment]
@beck I mean that behavior of npm's `link` and `install` commands a bit different. Unlike `install`, `link` doesn't trigger installation of package dependencies. So I don't see any problems with saving eslint plugins as "dependencies" for our config.

It may be ridiculous I did debug it on node v4.8.5 and v8.4.0 after I haven't found anything related to this in npm docs.
Attaching screenshots with v4.8.5:
Command `npm link eslint-config-yola` which is linked locally:
https://github.com/npm/npm/blob/release-next-2/lib/link.js#L106
https://github.com/npm/npm/blob/release-next-2/lib/build.js#L55
image

Command `npm install eslint-config-rusinov` which I published for test purposes:
https://github.com/npm/npm/blob/release-next-2/lib/install.js#L1163
https://github.com/npm/npm/blob/release-next-2/lib/build.js#L55
image

**Comment by RusinovAnton** [comment]
I would define react and jsx plugins once we decide on where to define our dependencies.

**Comment by RusinovAnton** [comment]
> This issue should be also solved when using npm version 3 which installs all subdependencies in the top-level node_modules folder.

So it isn't working for `npm < 3`

**Comment by RusinovAnton** [comment]
@beck generally, we need `npm@2` for lintreview only?

**Comment by RusinovAnton** [comment]
The workaround presented in [eslint-plugin-react-app](https://github.com/mmazzarolo/eslint-plugin-react-app/blob/2b0da6a2dcf53803a66d10fc22cfaf043786b609/index.js) isn't work for me on npm@2.15.11 as on npm@5.5.1
```
module.exports = {
  "extends": [
    "plugin:react-app/recommended"
  ],
  "rules": {
    "import/no-unresolved": "error"
  }
};
```

```
/home/rusinov/test-apps/test-lint-config/src/index.js
  1:1  error  Definition for rule 'import/no-unresolved' was not found  import/no-unresolved
```

**Comment by beck** [comment]
Adding here to document what I said on the call yesterday:
@RusinovAnton I believe you're seeing success with `npm install` because an NPM optimization is hiding the deficiency in eslint's module discovery.

The optimization is to try to flatten all dependencies. Not all dependencies are installed in the top-level `node_modules` but as many as possible. If all the child dependencies of `eslint-config-yola` and `eslint-config-airbnb` are installed in the top-level `node_modules` this is functionally equivalent to installing those dependencies as peer decencies.

Do not install a published version to verify you have resolved the dependency issues. Use `npm link` because this is the best way to expose the fault in eslint's module discovery.

Does ☝️  make sense?

**Comment by RusinovAnton** [comment]
@beck totally :100: . I've tested it with node@4 and npm@2 using package `npm ln`. And my approach introduced in current PR isn't worked as you said so I described my idea https://github.com/yola/production/issues/4137#issuecomment-341451847 and filed another PR
https://github.com/yola/eslint-plugin-yola/pull/1

**Comment by stefanor** [comment]
> npm@2

I don't think we have anything using npm 2 any more.

**Comment by beck** [comment]
> filed another PR

Just gave that PR a test drive locally, all looks good. I'm also seeing linting errors correctly displayed in my text editor.

Closing this in favor of: yola/eslint-plugin-yola#1
